Document register projections (#3752)
ansible/ansible-documentation: docs/docsite/rst/playbook_guide/playbooks_variables.rst

    - hosts: web_servers

      tasks:
        - name: Run a shell command and register multiple variables
          ansible.builtin.shell: /usr/bin/foo
          register:
            command_result: _task.result  # this is equivalent to register: command_result
            command_duration: _task.result.dt
            capitalized_command_output: _task.result.stdout | capitalize

Because this form of ``register`` is always a jinja expression, template delimiters ``{{}}`` are not required.

Variables registered in this way allow for chained access to other variables defined in the same ``register`` map. You can define variables based on other variables created in the same step, **and the order of definition does not matter**.

.. code-block:: yaml

    - hosts: web_servers

      tasks:
        - name: Run a shell command and register multiple variables
          ansible.builtin.shell: /usr/bin/foo
          register:
            capitalized_command_output: command_result.stdout | capitalize  # This works even though command_result is defined after capitalized_command_output
            command_result: _task.result  # this is equivalent to register: command_result

> Register projections are evaluated after the task completes, not during task execution. This means variables defined in a ``register`` map are not available within the task itself, and references to ``_task.result`` or other projected variables will reflect the final task state. This lazy evaluation enables order-independent variable definitions.

Author: ```suggestion
Register projections are evaluated after the task completes, not during task execution. For non-looped tasks, variables defined in a ``register`` map are not available within the task itself, and references to ``_task.result`` or other projected variables will reflect the final task state. This lazy evaluation enables order-independent variable definitions. For looped tasks, register projections from previous loop items are available, but require the ``default`` filter for the first iteration because ``_task.result`` is not yet defined.
```

> 

Reviewer: > Register projections are evaluated after the task completes, not during task execution. This means variables defined in a ``register`` map are not available within the task itself, and references to ``_task.result`` or other projected variables will reflect the final task state.

It might be good to clarify that this applies to non-looped tasks. Looped tasks could reference register projections in the task itself (would just need to use the `default` filter for the first iteration).

Author: Added a clarification.  I added another explicit clarification on *why* you need the `default` filter because it's probably one of the less obvious features.

For more examples of using registered variables in conditions on later tasks, see :ref:`playbooks_conditionals`. Registered variables may be simple variables, list variables, dictionary variables, or complex nested data structures. The documentation for each module includes a ``RETURN`` section that describes the return values for that module. To see the values for a particular task, run your playbook with ``-v``.

.. versionadded:: 2.21

You can also use ``register`` to register multiple variables and manipulate task output with jinja expressions with a dictionary of ``variable: expression`` pairs. Ansible provides an implicit task variable ``_task`` for accessing task output via its ``result`` property.

.. note::

>    ``_task`` is a reserved name and used for internal purposes. Do not use this name to register task results.

Reviewer: ```suggestion
   ``_task`` is a reserved name and used for internal purposes. Do not use this name to register task results or define variables.
```

We give runtime warnings for reserved variable names generally. It looks like trying to register _task is a fatal error (as expected), but if I define and reference a variable by this name it's silently ignored (this seems like a minor bug).

```yaml
- command: echo {{ _task }}
  vars:
    _task: silently_ignored
```

```console
changed: [localhost] => {
    "changed": true,
    "cmd": [
        "echo",
        "CurrentTask()"
    ],
    ...
```

.. code-block:: yaml+jinja

    - hosts: web_servers

      tasks:
        - name: Run a shell command and register multiple variables
          ansible.builtin.shell: /usr/bin/foo
          register:
            command_result: _task.result  # this is equivalent to register: command_result
            command_duration: _task.result.dt

>             capitalized_command_output: _task.result.stdout | capitalize

Reviewer: shell doesn't include dt in the result, so this example might be clearer if it calculates the duration from available keys

```suggestion
            command_duration: (command_end - command_start).total_seconds()
            command_start: _task.result.start | to_datetime(time_format)
            command_end: _task.result.end | to_datetime(time_format)
          vars:
            time_format: '%Y-%m-%d %H:%M:%S.%f'
```

> Because this form of ``register`` is always a jinja expression, template delimiters ``{{}}`` are not required. Do not use ``{{}}`` in the register projection expressions:

Reviewer 2: ```suggestion
Because this form of ``register`` is always a jinja expression, template delimiters ``{{ }}`` are not required. Do not use ``{{ }}`` in the register projection expressions:
```

Reviewer 2: I'm pretty sure the convention is to have a space in `{{ }}`.
